### PR TITLE
[nexus] Split Nexus configuration (package vs deployment)

### PR DESCRIPTION
**Pull request by smklein**
This is a major part of https://github.com/oxidecomputer/omicron/issues/822

This PR splits the Nexus configuration into "things known at package time" and "things known at deployment".

Due to hardcoded configurations, the two have historically been conflated. However, there are a number of
parameters which will only be known at deployment time, including:

- The addresses Nexus should use on the underlay (these are derived from the sled's `/64`, which is variable)
- The UUID of the Nexus instance (multiple Nexus instances may be executing simultaneously)
- The rack subnet
- (While it is still an IP address) the IP address of CRDB

### CODE REVIEW

**Comment by smklein** [comment]
For some background: 
image

File: nexus/src/config.rs [hunk 1]
```diff
@@ -124,22 +122,15 @@ impl Default for Tunables {
...
     /// Console-related tunables
     pub console: ConsoleConfig,
     /// Server-wide logging configuration.
     pub log: ConfigLogging,
-    /// Database parameters
-    pub database: db::Config,
     /// Authentication-related configuration
     pub authn: AuthnConfig,
     /// Timeseries database configuration.
+    // TODO: Should this be removed? Nexus needs to initialize it.
```
Comment (bnaecker) [inline nexus/src/config.rs:133]: This should probably be a runtime configuration setting, yes. I'm not sure whether that's nexus or RSS that'd be doing that. On the one hand, it might be nice to have some telemetry from early in the setup process. On the other, as it stands, `nexus` needs to be up (and CRDB) for `oximeter` to be collecting data from any clients. It's possible that some other component (not `oximeter`, maybe RSS?) could be writing data into ClickHouse during this phase, but I don't know how comfortable I am with that in general.

TL;DR, yes, but not sure who does that, probably Nexus.

File: nexus/src/config.rs [hunk 2]
```diff
@@ -150,74 +141,28 @@ pub struct Config {
...
 
-impl std::cmp::PartialEq<std::io::Error> for LoadError {
-    fn eq(&self, other: &std::io::Error) -> bool {
-        if let LoadErrorKind::Io(e) = &self.kind {
-            e.kind() == other.kind()
-        } else {
-            false
-        }
+impl Config {
+    /// Load a `PackageConfig` from the given TOML file
```
Comment (bnaecker) [inline nexus/src/config.rs:155]: Nit, but this is the top-level `Config`, not the `PackageConfig`. I guess everything is flattened at serialization, so maybe they're the same?

File: nexus/src/context.rs [hunk 3]
```diff
@@ -67,13 +68,13 @@ pub struct ConsoleConfig {
 impl ServerContext {
     /// Create a new context with the given rack id and log.  This creates the
     /// underlying nexus as well.
-    pub fn new(
+    pub async fn new(
```
Comment (bnaecker) [inline nexus/src/context.rs:71]: Did this really become `async`? I'm missing the `await` inside the body.

File: sled-agent/src/params.rs [hunk 4]
```diff
@@ -165,7 +163,7 @@ pub struct InstanceRuntimeStateRequested {
 pub enum DatasetKind {
     CockroachDb {
         /// The addresses of all nodes within the cluster.
-        all_addresses: Vec<SocketAddr>,
+        all_addresses: Vec<SocketAddrV6>,
```
Comment (bnaecker) [inline sled-agent/src/params.rs:166]: Yesssss

File: nexus/src/config.rs (hunk 1 above)
Comment (smklein) [reply nexus/src/config.rs:133]: Currently:

- Oximeter is launched by RSS, determined by hard-coded RSS configuration data

In the future:

- I hope Nexus manages Oximeter instances (provisioning them, making sure one is always up)
- ... and at that point, this argument probably goes away, because RSS shouldn't be doing the allocation anymore

File: nexus/src/config.rs (hunk 2 above)
Comment (smklein) [reply nexus/src/config.rs]: Comment updated

File: nexus/src/context.rs (hunk 3 above)
Comment (smklein) [reply nexus/src/context.rs]: It's a remnant from pulling it out of https://github.com/oxidecomputer/omicron/pull/1056 - it'll be necessary to use DNS services to look up where CRDB is running.

I'll remove it for now, however.

**Comment by smklein** [comment]
image

I went ahead and updated "runtime" to "deployment", as requested. Do ya'll have any other feedback, before I push forward on this?

File: nexus/src/config.rs (hunk 1 above)
Comment (smklein) [reply nexus/src/config.rs:135]: FYI this gets resolved in https://github.com/oxidecomputer/omicron/pull/1240 (Nexus can infer the name by DNS) and further in https://github.com/oxidecomputer/omicron/pull/1241 + https://github.com/oxidecomputer/omicron/pull/1216 , where Nexus provision oximeter / clickhouse wherever it wants.
